### ISSUE TITLE
DES3 192 bits key issue

### ISSUE THREAD

**Issue by uglym8** [OP-TEE/optee_os#3129]
Hi!

It seems that there's an issue in OP-TEE with usage of 192-bit DES3 keys.

I first encountered it in **3.1.0** and later verified that it is still present in the master as well.

I verified the problem existence by using [ DES3 example](https://github.com/uglym8/optee_examples/commits/des3_example) which is almost fully based on the [AES example](https://github.com/linaro-swg/optee_examples/tree/master/aes) .

I believe that these two checks (https://github.com/OP-TEE/optee_os/blob/master/core/tee/tee_svc_cryp.c#L1183 and https://github.com/OP-TEE/optee_os/blob/master/core/tee/tee_svc_cryp.c#L1187) are a little bit off for DES3 192-bit keys as the above mentioned comparisons want the keys to be:
1) evenly divisible by 56 (as we only use 7 bits of each byte)
2) be less than max size which is declared as 168 in https://github.com/OP-TEE/optee_os/blob/master/core/tee/tee_svc_cryp.c#L398

It seems that we should either change this to 192 bits or subtract 24 bits (for 192 bits keys) before these checks.

Otherwise this function will return TEE_ERROR_NOT_SUPPORTED which will be propagated to **syscall_cryp_obj_alloc()** making, for instance, DES3-CBC unusable with 192-bit keys.

I apologize if it is the false alarm but atm it seems  that it is not.

**Comment by vchong** [comment]
@uglym8

**NOTE**: It's summer time and many people in the core team is on vacation, so expect delayed answers to both questions as well as feedback in pull requests!

I took a look but am not quite sure about this. If you look at https://github.com/OP-TEE/optee_test/blob/master/host/xtest/regression_4000.c#L2352, the parity bits are taken out before calling `ta_crypt_cmd_allocate_operation()` and eventually `TEE_AllocateOperation()`, so if you do the same, your DES3 example will most probably work, but Section 6.2.1 of GPD_TEE_Internal_Core_API_Specification_v1.1.pdf states that the `maxKeySize` passed to `TEE_AllocateOperation()` should be based on Table 5-9 which, at least based on my interpretation, should include the parity bits. Perhaps others might interpret it differently? This seems to apply to the DES use cases as well.

Making the changes you suggested will make the 192-bit DES3 example work but break several other 400* series tests in `optee_test`, so we'll have to wait for others in the core team to return to confirm either the code or spec.

**Comment by uglym8** [comment]
> @uglym8
> 
> **NOTE**: It's summer time and many people in the core team is on vacation, so expect delayed answers to both questions as well as feedback in pull requests!
> 

Hey,
I am aware of that:)

> I took a look but am not quite sure about this. If you look at https://github.com/OP-TEE/optee_test/blob/master/host/xtest/regression_4000.c#L2352, the parity bits are taken out before calling `ta_crypt_cmd_allocate_operation()` and eventually `TEE_AllocateOperation()`,

Yes, that was I initially thought too.

> so if you do the same, your DES3 example will most probably work, but Section 6.2.1 of GPD_TEE_Internal_Core_API_Specification_v1.1.pdf states that the `maxKeySize` passed to `TEE_AllocateOperation()` should be based on Table 5-9 which, at least based on my interpretation, should include the parity bits. 
> 

Yes.
You are absolutely right. 

Having encountered this I even went one step further and tried to pass 21 as a key size:)Of course, this was rejected by OP-TEE.

> Making the changes you suggested will make the 192-bit DES3 example work but break several other 400* series tests in `optee_test`, so we'll have to wait for others in the core team to return to confirm either the code or spec.

I did not mean to suggest these as real changes otherwise I'd have created a PR instead.
I merely tried to show that at least changing them works but I did not assume that this could be a real fix.

**Comment by vchong** [comment]
> I did not mean to suggest these as real changes otherwise I'd have created a PR instead.
I merely tried to show that at least changing them works but I did not assume that this could be a real fix.

Ack. Just to be clear, I was trying to make xtest work by not removing the parity bits in https://github.com/OP-TEE/optee_test/blob/master/host/xtest/regression_4000.c#L2352 and either subtracting 24 bits before the 2 checks or changing the DES3 PROP to 64, 128, 192 in https://github.com/OP-TEE/optee_os/blob/master/core/tee/tee_svc_cryp.c#L398. This will allow both your example and xtest 4003 to pass, but caused 4002 and 4007 to fail : (, so the fix seemed a bit more involved than that.

If you'd like to take a shot at creating a PR for this, we'd most definitely welcome it. Otherwise, let's wait to see what the others have to say.

**Comment by jenswi-linaro** [comment]
I believe the checks are OK, but I'm open to discuss it of course.
There's no such thing as a 192 bit DES3 key unless you count the parity bits too.
We do have tests for 168 bit DES3 keys (192 bits with parity) in xtest:
https://github.com/OP-TEE/optee_test/blob/40aacb6dc33bbf6ee329f40274bfe7bb438bbf53/host/xtest/regression_4000.c#L2276-L2285

**Comment by uglym8** [comment]
@jenswi-linaro,

thanks for the quick update.

Of course, I was talking about 168 bits keys (w/o parity bits).
I'll verify the test this week and get back with results.

**Comment by uglym8** [comment]
@jenswi-linaro,

Just to make sure we're on the same page here: According to "Tee Internal Core API Specification - Public Review v1.1.2.50 (Target v1.2) Table 5-9 (page 2494) under "Possible Key Sizes" one can clearly see that acceptable key size for, say, 3DES are 128 and 192 bits **including the parity bits**.

Actually @vchong also mentioned this as well.

Having read this I made a conclusion that I should use either 168 or 192 as a key size parameter for TEE_AllocateOperation(). 
But the cases you mentioned in your reply actually modify the key size before calling TEE_AllocateOperation().

So far the question is should I use key size with parity bits or without?

P.S. I'm still investigating the problem so more likely that there will be some more information.

**Comment by jenswi-linaro** [comment]
@uglym8, I agree that this is confusing. I'm a bit surprised that the specification isn't more clear on this point since we had confusion on this when working with the conformance test suites on v 1.0.

However, OP-TEE passes the DES* tests in the GP test suite as far as I know so the GP tests must be passing the DES* key size without parity just as xtest does.

**Comment by jbech-linaro** [comment]
@jforissier , do you think you can ask the GP guys to get a clarification?

**Comment by jforissier** [comment]
@jbech-linaro not sure ;-) anyway, I think I have a pretty clear view of the situation now.

**TL;DR**

OP-TEE is doing the right thing, which is:
- It expects the "raw" DES/3DES key size, **not** counting the parity bits, in `TEE_AllocateOperation()` and `TEE_AllocateTransientObject()`. For example: 168.
- It expects the keys are passed to `TEE_PopulateTransientObject()` **with** parity bits in the **LSB of each byte**, thus requiring a larger buffer than the size passed to the allocation functions. For example, for a 'raw' 3DES key size of 168 as above, that means supplying a buffer of 192 bits (3 * 56 bits of key material is 3 * (7 * 8), add a parity bit to each 7-bit quantity and you get 3 * ((7 + 1) * 8) = 3 * 64 = 192).

**More details**

- About the actual format of the DES/3DES key buffer passed to `TEE_PopulateTransientObject()`
  - If you try randomly flipping the LSB of each byte in xtest's [ciph_data_des3_key1](https://github.com/OP-TEE/optee_test/blob/40aacb6dc33bbf6ee329f40274bfe7bb438bbf53/host/xtest/regression_4000.c#L1730), you will see that the test still passes (`xtest 4003`).
  - This proves that this LSB is not used as key material (and is not checked either by OP-TEE... which is arguably not a good thing)
- About the GP spec (TEE Internal [Core] API)
  - Version 1.0 has "112 or 168 bits" for "Possible object size" of "Object Type TEE_TYPE_DES3" in table 5-7.
  - Version 1.1 and 1.2 have the confusing text "128 or 192 bits including the parity bits. This gives effective key sizes of 112 or 168 bits" at the same location (that's table 5-9).
  - The intent of adding the text in v1.1 is clarified in a GP document called "GPD_TEE_Internal_API_v1.0_EP_v3.0.pdf", *Errata and Precisions v3.0* (GPD_EPR_017), which says:

    > P.6 Clarify the Object Size to Pass to Allocate a Key Object
    >
    > The specification needs a clear statement that the value to be passed to allocate a key object
    > is related to the strength of the related key. This is particularly needed for DES and 3 DES
    > algorithms as keys related to those algorithms are longer than their strength due to parity bits.

    The EP document then goes on with proposed changes that were partly included in v1.1 and later. Unfortunately, the new text remains quite a bit ambiguous as we can see, it's unfortunate that GP chose not to mention "cryptographic strength" for example.

- I realize that there is no DES/DES3 test in the GP test suite we are using with `xtest`!
